### PR TITLE
Introduced common functions to remove code duplication in mechanics related processes

### PR DESCRIPTION
**Pull request by wenqing**
As titled.
The introduced common functions are included in a new file: 
`ProcessLib/Utils/SetOrGetIntegrationPointData.h`.

### CODE REVIEW

File: ProcessLib/Utils/SetOrGetIntegrationPointData.h [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        ip_data[ip].sigma_eff =
+            MathLib::KelvinVector::symmetricTensorToKelvinVector(
+                sigma_values.col(ip));
+    }
+
+    return num_intpts;
+}
+
+template <typename IP_Data, typename IP_DataMember>
+std::vector<double> const& getIntegrationPintScalarData(
```
Comment (wenqing) [inline ProcessLib/Utils/SetOrGetIntegrationPointData.h:113]: This function may increase a little bit compilation time.

Comment (endJunction) [reply ProcessLib/Utils/SetOrGetIntegrationPointData.h]: How much?

**Comment by endJunction** [comment]
There are a lot of changes due to formatting. Please extract them to a single commit if the changes are necessary. Difficult to review otherwise.

File: ProcessLib/Utils/SetOrGetIntegrationPointData.h (hunk 1 above)
Comment (endJunction) [reply ProcessLib/Utils/SetOrGetIntegrationPointData.h]: Typo

File: ProcessLib/SmallDeformation/SmallDeformationFEM.h [hunk 2]
```diff
@@ -47,7 +48,10 @@ struct IntegrationPointData final
     {
     }
 
-    typename BMatricesType::KelvinVectorType sigma, sigma_prev;
+    // For the consistency with the name of other mechanical process related
+    // processes,
+    // sigma is renamed as sigma_eff
+    typename BMatricesType::KelvinVectorType sigma_eff, sigma_prev;
```
Comment (endJunction) [inline ProcessLib/SmallDeformation/SmallDeformationFEM.h:54]: The comment goes into the commit message.
sigma_prev must be renamed too.

File: ProcessLib/SmallDeformation/SmallDeformationFEM.h [hunk 3]
```diff
@@ -288,7 +297,7 @@ class SmallDeformationLocalAssembler
             auto const& sigma_prev = _ip_data[ip].sigma_prev;
 
             auto& eps = _ip_data[ip].eps;
-            auto& sigma = _ip_data[ip].sigma;
+            auto& sigma = _ip_data[ip].sigma_eff;
```
Comment (endJunction) [inline ProcessLib/SmallDeformation/SmallDeformationFEM.h:300]: also here the rename is missing. Please rename all of the sigmas.

File: ProcessLib/RichardsMechanics/RichardsMechanicsFEM-impl.h [hunk 4]
```diff
@@ -1133,17 +1108,8 @@ std::vector<double> const& RichardsMechanicsLocalAssembler<
...
-    auto cache_mat = MathLib::createZeroedMatrix<
-        Eigen::Matrix<double, 1, Eigen::Dynamic, Eigen::RowMajor>>(cache, 1,
-                                                                   num_intpts);
-
-    for (unsigned ip = 0; ip < num_intpts; ++ip)
-    {
-        cache_mat[ip] = _ip_data[ip].saturation;
-    }
+    ProcessLib::getIntegrationPointScalarData(_ip_data, &IpData::saturation,
+                                             cache);
```
Comment (TomFischer) [inline ProcessLib/RichardsMechanics/RichardsMechanicsFEM-impl.h:1112]:
```suggestion
                                              cache);
```

File: ProcessLib/RichardsMechanics/RichardsMechanicsFEM-impl.h [hunk 5]
```diff
@@ -1170,18 +1136,8 @@ std::vector<double> const& RichardsMechanicsLocalAssembler<
...
-        Eigen::Matrix<double, 1, Eigen::Dynamic, Eigen::RowMajor>>(cache, 1,
-                                                                   num_intpts);
-
-    for (unsigned ip = 0; ip < num_intpts; ++ip)
-    {
-        cache_mat[ip] = _ip_data[ip].porosity;
-    }
-
+    ProcessLib::getIntegrationPointScalarData(_ip_data, &IpData::porosity,
+                                             cache);
```
Comment (TomFischer) [inline ProcessLib/RichardsMechanics/RichardsMechanicsFEM-impl.h:1140]:
```suggestion
                                              cache);
```

File: ProcessLib/SmallDeformationNonlocal/SmallDeformationNonlocalFEM.h [hunk 6]
```diff
@@ -689,71 +680,21 @@ class SmallDeformationNonlocalLocalAssembler
...
-
-        for (unsigned ip = 0; ip < n_integration_points; ++ip)
-        {
-            _ip_data[ip].sigma =
-                MathLib::KelvinVector::symmetricTensorToKelvinVector(
-                    sigma_values.col(ip));
-        }
-
-        return n_integration_points;
+        return ProcessLib::setSigma<IP_Data, DisplacementDim>(values, _ip_data);
```
Comment (endJunction) [inline ProcessLib/SmallDeformationNonlocal/SmallDeformationNonlocalFEM.h:689]: No underscores in type names.
https://google.github.io/styleguide/cppguide.html#Type_Names

File: ProcessLib/SmallDeformation/SmallDeformationFEM.h [hunk 7]
```diff
@@ -47,7 +48,10 @@ struct IntegrationPointData final
     {
     }
 
-    typename BMatricesType::KelvinVectorType sigma, sigma_prev;
+    // For the consistency with the name of other mechanical process related
+    // processes,
+    // sigma is renamed as sigma_eff
```
Comment (TomFischer) [inline ProcessLib/SmallDeformation/SmallDeformationFEM.h:53]:
```suggestion
    // For the consistency with the name of other mechanical process related
    // processes, sigma is renamed as sigma_eff
```

File: ProcessLib/SmallDeformationNonlocal/IntegrationPointData.h [hunk 8]
```diff
@@ -37,7 +37,10 @@ struct IntegrationPointData final : public IntegrationPointDataNonlocalInterface
     }
 
     typename BMatricesType::BMatrixType b_matrices;
-    typename BMatricesType::KelvinVectorType sigma, sigma_prev;
+    // For the consistency with the name of other mechanical process related
+    // processes,
+    // sigma is renamed as sigma_eff
```
Comment (TomFischer) [inline ProcessLib/SmallDeformationNonlocal/IntegrationPointData.h:42]: Same as above.

File: ProcessLib/ThermoMechanics/ThermoMechanicsFEM.h [hunk 9]
```diff
@@ -45,7 +45,10 @@ struct IntegrationPointData final
     }
 
     // total stress
-    typename BMatricesType::KelvinVectorType sigma, sigma_prev;
+    // For the consistency with the name of other mechanical process related
+    // processes,
+    // sigma is renamed as sigma_eff
```
Comment (TomFischer) [inline ProcessLib/ThermoMechanics/ThermoMechanicsFEM.h:50]: Again, same as above.

File: ProcessLib/Utils/SetOrGetIntegrationPointData.h [hunk 10]
```diff
@@ -0,0 +1,132 @@
+/**
+ * \file
+ * \copyright
+ * Copyright (c) 2012-2020, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ * Created on May 4, 2020, 10:27 AM
```
Comment (TomFischer) [inline ProcessLib/Utils/SetOrGetIntegrationPointData.h:9]: This is not necessary - the information is stored in the commit.

File: ProcessLib/Utils/SetOrGetIntegrationPointData.h [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+    auto const num_intpts = ip_data.size();
+
+    auto sigma_values =
+        Eigen::Map<Eigen::Matrix<double, kelvin_vector_size, Eigen::Dynamic,
+                                 Eigen::ColMajor> const>(
+            values, kelvin_vector_size, num_intpts);
+
+    for (unsigned ip = 0; ip < num_intpts; ++ip)
+    {
+        ip_data[ip].sigma_eff =
```
Comment (endJunction) [inline ProcessLib/Utils/SetOrGetIntegrationPointData.h:104]: Why not pass a pointer to member here as for the scalar data?
Would avoid all of the renames too and make the distinction between the epsilon and sigma unnecessary.
Same as in https://github.com/ufz/ogs/blob/master/ProcessLib/RichardsMechanics/RichardsMechanicsFEM-impl.h#L889-L913

File: ProcessLib/Utils/SetOrGetIntegrationPointData.h (hunk 1 above)
Comment (wenqing) [reply ProcessLib/Utils/SetOrGetIntegrationPointData.h]: That was just an estimation before. Actually there is no change in the build time due to this PR:
**Comparison the times of  full building in release on my laptop:**
```
ufz/master:        ninja -j4  4293.27s user 97.02s system 394% cpu 18:31.89 total
Branch of this PR: ninja -j4  4288.14s user 97.04s system 394% cpu 18:31.43 total
```

File: ProcessLib/SmallDeformation/SmallDeformationFEM.h (hunk 2 above)
Comment (wenqing) [reply ProcessLib/SmallDeformation/SmallDeformationFEM.h]: No sun change anymore by passing class member via function argument.

File: ProcessLib/RichardsMechanics/RichardsMechanicsFEM-impl.h (hunk 4 above)
Comment (wenqing) [reply ProcessLib/RichardsMechanics/RichardsMechanicsFEM-impl.h]: Formatted

File: ProcessLib/RichardsMechanics/RichardsMechanicsFEM-impl.h (hunk 5 above)
Comment (wenqing) [reply ProcessLib/RichardsMechanics/RichardsMechanicsFEM-impl.h]: Formatted

File: ProcessLib/SmallDeformationNonlocal/SmallDeformationNonlocalFEM.h (hunk 6 above)
Comment (wenqing) [reply ProcessLib/SmallDeformationNonlocal/SmallDeformationNonlocalFEM.h]: Changed.

File: ProcessLib/SmallDeformation/SmallDeformationFEM.h (hunk 7 above)
Comment (wenqing) [reply ProcessLib/SmallDeformation/SmallDeformationFEM.h]: It is gone.

File: ProcessLib/PhaseField/PhaseFieldFEM.h [hunk 12]
```diff
@@ -283,24 +286,9 @@ class PhaseFieldLocalAssembler : public PhaseFieldLocalAssemblerInterface
...
-        for (unsigned ip = 0; ip < num_intpts; ++ip)
-        {
-            auto const& sigma = _ip_data[ip].sigma;
-            cache_mat.col(ip) =
-                MathLib::KelvinVector::kelvinVectorToSymmetricTensor(sigma);
-        }
-
-        return cache;
+        return ProcessLib::getIntPtKelvinVector<
+            IpData, decltype(&IpData::sigma), DisplacementDim>(
```
Comment (endJunction) [inline ProcessLib/PhaseField/PhaseFieldFEM.h:290]: what is the reason for the `decltype`?

File: ProcessLib/Utils/SetOrGetIntegrationPointData.h (hunk 11 above)
Comment (wenqing) [reply ProcessLib/Utils/SetOrGetIntegrationPointData.h]: Beside, for standalone function, it needs to set member type in the template parameter if  template parameters have non-type ones (such as `int DispacementDim`).

File: ProcessLib/Utils/SetOrGetIntegrationPointData.h (hunk 10 above)
Comment (wenqing) [reply ProcessLib/Utils/SetOrGetIntegrationPointData.h:9]: It records the moment of a mouse click to create the file (not the time at commit), and displays the time in documentation. If you strongly object it, I will remove it.

File: ProcessLib/SmallDeformation/SmallDeformationFEM.h [hunk 13]
```diff
@@ -384,49 +387,19 @@ class SmallDeformationLocalAssembler
...
-        {
-            auto const& sigma = _ip_data[ip].sigma;
-            cache_mat.row(ip) =
-                MathLib::KelvinVector::kelvinVectorToSymmetricTensor(sigma);
-        }
-
-        return ip_sigma_values;
+        return ProcessLib::getKelvinVectorValues<
+            IpData, decltype(&IpData::sigma), DisplacementDim>(_ip_data,
+                                                               &IpData::sigma);
```
Comment (endJunction) [inline ProcessLib/SmallDeformation/SmallDeformationFEM.h:402]: I think, if you reorder the template arguments of `getKelvinVectorValues` to`
```c++
template <int DisplacementDim, typename IpData, typename Member>
std::vector<double> getKelvinVectorValues(IpData const&, Member const&);
```
then you can omit the `IpData` and the member pointer types.

Comment (wenqing) [reply ProcessLib/SmallDeformation/SmallDeformationFEM.h:402]: Going to try your suggestion.

Comment (wenqing) [reply ProcessLib/SmallDeformation/SmallDeformationFEM.h]: It works.

File: ProcessLib/Utils/SetOrGetIntegrationPointData.h [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+        ip_data[ip].*member =
+            MathLib::KelvinVector::symmetricTensorToKelvinVector(
+                kelvin_vector_values.col(ip));
+    }
+
+    return num_intpts;
+}
+
+template <typename IntegrationPointData, typename MemberType>
+std::vector<double> const& getIntegrationPointScalarData(
```
[the remark this answers is not in the document]
Comment (wenqing) [reply ProcessLib/Utils/SetOrGetIntegrationPointData.h:102]: Changed.

You can just review the new commits (by commit).

[the remark this answers is not in the document]
Comment (endJunction) [reply ProcessLib/Utils/SetOrGetIntegrationPointData.h:102]: First two do look good.
The last one with `cacheSomething` is not good, in my opinion.

[the remark this answers is not in the document]
Comment (endJunction) [reply ProcessLib/Utils/SetOrGetIntegrationPointData.h:102]: The `cache` for return type should be removed on the long run anyway.

[the remark this answers is not in the document]
Comment (wenqing) [reply ProcessLib/Utils/SetOrGetIntegrationPointData.h:102]: How about `getIntegrationPointKelvinVectorWithCache?

[the remark this answers is not in the document]
Comment (endJunction) [reply ProcessLib/Utils/SetOrGetIntegrationPointData.h:102]: Why not keep it as at was before: `getIntegrationPointKelvinVector`?

**Comment by wenqing** [comment]
Any further comment?

File: ProcessLib/Utils/SetOrGetIntegrationPointData.h (hunk 14 above)
[the remark this answers is not in the document]
Comment (wenqing) [reply ProcessLib/Utils/SetOrGetIntegrationPointData.h:102]: Changed.

**Comment by endJunction** [comment]
@wenqing I'd like to squash it, unless you somehow provide concise commit messages.

**Comment by wenqing** [comment]
Merged some commits, especially those for replying the review comments.

**Comment by endJunction** [comment]
@wenqing Clang-formatted the commits: please do it yourself next time.
One more commit for consistent naming of the functions.
If you agree and tests don't turn red all over the place, can be merged.

**Comment by wenqing** [comment]
@endJunction The changed name by a function overloading is OK.

**Pull request merged**